### PR TITLE
fix a bug in add uncoalesced tensors and fully test sparse math

### PR DESCRIPTION
**Pull request by nairbv**
Dividing an uncoalesced integral tensor doesn't give accurate results:

for a sparse int tensor with value 6 represented as values=[3, 3], integer division by 2 would give values=[1, 1] => 2 instead of 6 / 2 => 3. To fix this, we need to coalesce the tensors.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30318 add type promotion support for sparse tensors
* **#30317 fix a bug in add uncoalesced tensors and fully test sparse math**

### CODE REVIEW

**Comment by zou3519** [comment]
What exactly is wrong with uncoalesced addition? It would be helpful if we had a problem statement "uncolesced addition is incorrect because of X" and a solution statement "we fixed it by doing W, Y, and Z" -- as a reviewer I am not sure what I am looking at.

**Comment by nairbv** [comment]
> What exactly is wrong with uncoalesced addition? It would be helpful if we had a problem statement "uncolesced addition is incorrect because of X" and a solution statement "we fixed it by doing W, Y, and Z" -- as a reviewer I am not sure what I am looking at.

Updated the PR summary to include the explanation from the comments.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 1]
```diff
@@ -169,8 +180,17 @@ SparseTensor& div_out_sparse_zerodim(SparseTensor& r, const SparseTensor& t, con
...
   if (is_same_tensor(r, t)) {
+    // Can't divide an uncoalesced integral tensor accurately. e.g. for a sparse int tensor with value 6
+    // represented as values=[3, 3], integer division by 2 would give values=[1, 1] => 2 instead
+    // of 6 / 2 => 3
+    if (!r.is_coalesced() && isIntegralType(r.scalar_type(), true)) {
+      coalesce_(r);
+    }
     r._values().div_(value);
   } else {
+    if (!t.is_coalesced() && isIntegralType(r.scalar_type(), true)) {
```
Comment (zou3519) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp]: Right now in the code there are two coalesce checks and calls to in-place coalesce. We can move this to before checking `if_same_tensor(r, t)` so that we only have to check that `t` is coalesced in one location

Furthermore: shouldn't this be a `isIntegralType(t.scalar_type(), true)` (t instead of r), especially since there isn't type promotion in this patch yet?

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 2]
```diff
@@ -161,6 +161,17 @@ Tensor& div_sparse_(Tensor& self, const Tensor& value) {
   return div_out_sparse_zerodim(self, self, value);
 }
 
+// inplace coalesce
+const SparseTensor& coalesce_(const SparseTensor& tensor) {
+  SparseTensor coalesced = tensor.coalesce();
```
Comment (zou3519) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp]: In-place sparse division doesn't actually save any memory because we are coalescing out-of-place here (the point of in-place operations is to save on memory usage). This would be good to document in the sparse docs

File: test/test_type_promotion.py [hunk 3]
```diff
@@ -172,18 +172,19 @@ def test_mixed_type_backward(self, device):
                 func(x, y).sum().backward()
 
     def _get_test_tensor(self, device, dtype, remove_zeros=False):
-        shape = [20, 20, 20]
+        shape = [10, 10, 10]
```
Comment (zou3519) [inline test/test_type_promotion.py]: Why did this change?

File: test/test_type_promotion.py [hunk 4]
```diff
@@ -430,6 +431,97 @@ def test_transpose(self, device):
         a = torch.tensor([[True, True], [False, True]], device=device)
         self.assertEqual(a.t() == 0, a.t() == False)  # noqa: E712
 
+    def get_sparse_tensors(self, device, dtype, coalesced, zeros=True):
+        t = self._get_test_tensor(device, dtype, not zeros)
+        if coalesced:
+            s = t.to_sparse()
```
Comment (zou3519) [inline test/test_type_promotion.py]: It would be good to assert s is coalesced

File: test/test_type_promotion.py [hunk 5]
```diff
@@ -430,6 +431,97 @@ def test_transpose(self, device):
...
+        else:
+            s = t.to_sparse()
+            indices = torch.cat((s.indices(), s.indices()), 1)
+            values = torch.cat((s.values(), s.values()), 0)
+            s = torch.sparse_coo_tensor(indices=indices, values=values, size=s.size(), dtype=dtype, device=device)
+            t = s.to_dense()
```
Comment (zou3519) [inline test/test_type_promotion.py]: Would be good to assert s isn't coalesced

File: test/test_type_promotion.py [hunk 6]
```diff
@@ -430,6 +431,97 @@ def test_transpose(self, device):
...
+        return (t, s)
```
Comment (zou3519) [inline test/test_type_promotion.py]: We're returning a dense tensor (t) from a function called "get_sparse_tensors"?

**Comment by zou3519** [comment]
So nothing is wrong with sparse addition, the bug is with sparse division? The title of the PR seems wrong

File: test/test_type_promotion.py [hunk 7]
```diff
@@ -430,6 +431,97 @@ def test_transpose(self, device):
...
+
+    def _get_precision(self, dtype, coalesced):
+        if dtype == torch.half and not coalesced:
+            # very low precision for uncoalesced float16 sparse tensors since
```
Comment (zou3519) [inline test/test_type_promotion.py]: This seems bad

File: test/test_type_promotion.py [hunk 8]
```diff
@@ -430,6 +431,97 @@ def test_transpose(self, device):
...
+        if op_name != 'div':
+            sparse = op(s1, s2)
+            self.assertEqual(sparse.dtype, e.dtype)
+            self.assertEqual(e, sparse.to_dense(), prec=precision, message=err)
+        else:
+            # sparse division only supports division by a scalar
+            self.assertRaises(RuntimeError, lambda: op(s1, s2).to_dense())
+
+        # Test op(dense, sparse)
+        if add_sub:
```
Comment (zou3519) [inline test/test_type_promotion.py]: There's a lot of special casing going on in this function to handle testing for 4 operators: add, sub, mul, div. The special casing makes it so that it is hard to tell whether or not the code is comprehensively testing each operator. It might be better to break it up into 4 tests "test_add",  "test_sub", "test_mul", "test_div".

Furthermore, these tests should exist in `test_sparse` because they don't do any type promotion

**Review by zou3519 — CHANGES_REQUESTED** [review]
initial pass through. The fix looks right but the sparse testing is a bit hard to read

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp (hunk 2 above)
Comment (nairbv) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: right, the purpose here is just to be able to have the change affect the reference. My intent wasn't to expose this to users.

Comment (zou3519) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: My suggestion was not to document "coalesce_" in the docs; it is to document to users that "in-place sparse division doesn't save any memory in its current implementation so there is no reason to use it over regular sparse division"

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp (hunk 1 above)
Comment (nairbv) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: If we move it before the `if (is_same_tensor..` then we have to always coalesce `t` and not `r`. Then the reader has to understand that in the case where we return `r` we're relying on having in-place coalesced an alias of it. It's less code I guess and I suppose that also means the comment is more targeted but are you confident that's less confusing to read?

> shouldn't this be a...

It's actually still `r`. In most cases r.scalar_type==t.scalar_type, but in the cases where it doesn't it would be because someone passed r as `out=`. In that case, if r is floating, we probably don't need to take the perf hit of the in-place coalesce.
```
>>> s=torch.ones([3,3], dtype=torch.int).to_sparse()
>>> t=torch.ones([3,3]).to_sparse()
>>> torch.div(s, 2, out=t)
tensor(indices=tensor([[0, 0, 0, 1, 1, 1, 2, 2, 2],
                       [0, 1, 2, 0, 1, 2, 0, 1, 2]]),
       values=tensor([0.5000, 0.5000, 0.5000, 0.5000, 0.5000, 0.5000, 0.5000,
                      0.5000, 0.5000]),
       size=(3, 3), nnz=9, layout=torch.sparse_coo)
```

File: test/test_type_promotion.py (hunk 3 above)
Comment (nairbv) [reply test/test_type_promotion.py]: I could change it back, but it just seemed excessive and we have a lot of tests to run.

File: test/test_type_promotion.py (hunk 7 above)
Comment (nairbv) [reply test/test_type_promotion.py]: yeah that's just a fact of low precision though:

```
In [20]: values=[10.98764] * 4
In [21]: values
Out[21]: [10.98764, 10.98764, 10.98764, 10.98764]
In [22]: torch.tensor(values, dtype=torch.half, device='cuda').sum() - torch.tensor(values, dtype=torch.float).sum()
Out[22]: tensor(-0.0131, device='cuda:0')
```

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: > It's actually still r. In most cases r.scalar_type==t.scalar_type, but in the cases where it doesn't it would be because someone passed r as out=. In that case, if r is floating, we probably don't need to take the perf hit of the in-place coalesce.

Oh, I see, thanks for explaining that. I didn't realize we converted values() to the correct type before division if the out tensor has another type.

Comment (zou3519) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: > If we move it before the if (is_same_tensor.. then we have to always coalesce t and not r. Then the reader has to understand that in the case where we return r we're relying on having in-place coalesced an alias of it. 

The case that you're describing is the in-place case where `tensor.div_(foo)` calls `div_out(tensor, tensor, foo)`. This is a common paradigm in our codebase: when we implement an in-place function and there already exists an out variant, we call the out-variant with `result` and `input` equal to the tensor, so I don't think this makes the code harder to read.

File: test/test_type_promotion.py (hunk 8 above)
Comment (nairbv) [reply test/test_type_promotion.py]: I think this actually makes it easier to see that every case is captured, since every `else` is an `assertRaises`. We're negatively affirming all of the uncaptured conditions, which ensures that every supported possibility is tested. If I use distinct test_mul/test_div functions, it's easy to forget one negative assertion for something like a mul(sparse,sparse) case. I'd like to be able to enforce that as we improve sparse support, we are forced to fix the tests.

I'm also afraid that if I break them apart into separate functions I'll either be re-constructing cuda tensors for each operator (which could be slow) or passing around too many parameters for it to be clearly readable.

To preserve the negative tests, if you really think it would be more readable even with many parameters being passed to each test, I'd probably prefer to break it up by dense_sparse, sparse_sparse, etc, instead of by operator.

In other words, this section would look something like:
```
test_sparse_sparse(op_name, inplace, s1, s2, e)
test_dense_sparse(op_name, inplace, d1, s2, e)
test_sparse_scalar(....)
```
the implementations would probably end up repeating some code like the `suffix = '_' if inplace else ''` above.

Or I could just make them nested functions that close over the variables in the setup part, if you think that would be clearer?

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp (hunk 2 above)
Comment (nairbv) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: ah OK, though we only use extra memory in the case where the input tensor is an uncoalesced integral tensor. That seems a bit narrow to be explaining the implementation detail.

File: test/test_type_promotion.py (hunk 8 above)
Comment (zou3519) [reply test/test_type_promotion.py]: Here's one example of the things that seem weird to me. We call:
```
dtypes = torch.testing.get_all_math_dtypes(device)
        ops = ['add', 'sub', 'mul', 'div']
        for dtype in dtypes:
            for op_name in ops:
                for inplace, coalesced in itertools.product([True, False], [True, False]):
                    self._test_sparse_op(op_name, inplace, dtype, device, coalesced)
```
Okay, this seems reasonable. I would expect that we are testing the same thing for each operator. 

It turns out, that expectation isn't true. There's a special case for division where we are only testing division by a scalar but not division of two sparse tensors. If someone comes along in the future and adds support for division of two sparse tensors, where do they write a new test case? They would have to:
1) know about the existence of this test (it isn't clear from the test name that this is testing sparse division)
2) read through this function to find where to enable the test for sparse/sparse division.

If, instead, we had a `test_div_sparse_sparse(self)` test case, then a future developer could just add their test case in there and not have to worry about understanding how `test_sparse_ops` is doing all of its checks.

Comment (nairbv) [reply test/test_type_promotion.py]: > There's a special case for division where we are only testing division by a scalar but not division of two sparse tensors.

We are though, we're explicitly testing the current behavior of division of two sparse tensors, where the current behavior is to raise an error. If someone changes that behavior (by implementing sparse division), they'll be forced to change the test (since it'll start failing when it no longer raises an error). That's a testing style @gchanan often requests for unsupported operations.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: [is_coalesced](https://pytorch.org/docs/stable/sparse.html?highlight=coalesce#torch.sparse.FloatTensor.is_coalesced) is a public property and it's very easy to get an uncoalesced tensor when working with sparse tensors (i.e., sparse addition). I wouldn't call this an implementation detail; it is part of the API that an in-place operation saves memory. 

It is important for us to note places where the behavior might differ from what the user is expecting, either by documenting it correctly or filing an issue to be fixed.

File: test/test_type_promotion.py (hunk 3 above)
Comment (zou3519) [reply test/test_type_promotion.py]: You don't need to change it back, I was just curious

File: test/test_type_promotion.py (hunk 8 above)
Comment (zou3519) [reply test/test_type_promotion.py]: I feel like we're discussing a few points at once, so let's just go through them one by one.

First, could we move this test case to test_sparse.py? It doesn't relate to type promotion and all the other tests for binary functions like this exist in test_sparse.

Comment (nairbv) [reply test/test_type_promotion.py]: > First, could we move this test case to test_sparse.py

It will be testing mixed-types though in the next diff. When I initially wrote this test, I actually skipped same-type combinations, and only re-added them for this diff when I came across this bug. Is it really worth moving now just to move back in the next diff, or would you want it to stay in test_sparse even when it is updated to test mixed-types?

Comment (zou3519) [reply test/test_type_promotion.py]: Since this diff is fixing uncoalesced division, it makes more sense to me to add a single test case for uncoalesced division into test_sparse, and move all of the type promotion testing code into the next diff.

Comment (zou3519) [reply test/test_type_promotion.py]: I tried to write up what I think would be a clearer alternative to the test code here, but it doesn't save any LOC and also has its tradeoffs :/ https://gist.github.com/zou3519/3c6248a7e2267c796043b6ef59561981 . 

I don't know if it is better, but this is what I mean by wanting something clearer that doesn't have special casing:
- there is a clear `test` function just does what it is supposed to (get test inputs, run the test, exit)
- the table has what is supported, what isn't supported, and which test cases need inputs without zeros.

Comment (nairbv) [reply test/test_type_promotion.py]: > Since this diff is fixing uncoalesced division

I'd generally favor readability of code as-is over readability of the diff. I think if we can create a set of test that completely covers all cases, it's unhelpful to retain the narrow individual cases that overlap with the broader tests. Statically, the narrow tests will only really be useful to someone looking back at old diffs/issues, and doesn't say much about the current state of the code as it exists fully merged without the bugs present.

I'd actually like to get rid of most of the tests in test_type_promotion that overlap with test_many_promotions, at least to the extent that those tests are testing operations that are fully tested by comprehensive tests. Some older tests like `test_int_promotion` that only existed due to particular examples in historical discussions aren't really helpful to understanding the current state of the code as it exists now. If I landed the narrow test just to check the one specific bug I discovered with these tests, I'd kind of prefer to remove that no-longer-needed test once we had complete coverage in the next diff anyway.

> I tried to write up ...

I find this a bit confusing, so I think this might just be a stylistic difference. My main concern with this approach though would be having the list of tests hard-coded, since it leaves a lot of room for missing a case. There's a fair bit that would also still need to be added too (e.g. can_cast errors in the next diff are dependent on the dtypes of the operands), so I think this would end up being just as much code. Plus by being specific the comments can describe each particular case in the right place, and by re-using test tensors created for out of place operations, we save a bit of run-time over creating new test tensors for each test.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp (hunk 2 above)
Comment (nairbv) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: Is there somewhere in the docs that already goes into this level of detail about the memory tradeoffs of in-place ops and impact of uncoalesced tensors? It seems like it would be odd to document to this level of detail just for this particular change without the rest of the context you refer to. If adding the rest of the context, I'd prefer to use a separate PR. 

The only explanation I could find was the note here:
https://pytorch.org/docs/stable/sparse.html
which is pretty general, "For the most part, you shouldn’t have to care whether or not a sparse tensor is coalesced or not..."

Comment (zou3519) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: > Is there somewhere in the docs that already goes into this level of detail about the memory tradeoffs of in-place ops and impact of uncoalesced tensors? 

It looks like we don't document the semantics of "in-place" operations. However, users use those as a strategy to save memory.

See
[this](https://towardsdatascience.com/in-place-operations-in-pytorch-f91d493e970e), and [this](https://discuss.pytorch.org/t/what-is-in-place-operation/16244) (a few things I found from a quick search)

> The only explanation I could find was the note here:
https://pytorch.org/docs/stable/sparse.html
which is pretty general, "For the most part, you shouldn’t have to care whether or not a sparse tensor is coalesced or not.."

I don't know if `div_out` saves memory. If it does, and if we are adding a new case where it doesn't save memory, then saying that a "user shouldn't care whether or not a sparse tensor is coalesced" means that we should not be inconsistent over whether or not `div_out` saves memory or not.

File: test/test_type_promotion.py (hunk 6 above)
Comment (nairbv) [reply test/test_type_promotion.py]: renamed

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp (hunk 2 above)
Comment (nairbv) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: > I don't know if div_out saves memory.

It saves memory in the particular case of in-place division (e.g. sparse_tensor /= 2.0) on coalesced or floating tensors, but I don't see the general characteristics being a documented anywhere yet (i.e. this change doesn't make any existing documentation wrong). I'm interpreting "user shouldn't care" as meaning it doesn't change the semantics of their code -- obviously they do sometimes already care since it does impact performance (regardless of this current PR).

It could be helpful to document the memory characteristics of sparse and uncoalesced ops in more detail than we have, but that seems outside the scope of this PR.

I'm not entirely convinced that we should go into too much detail when/if we do, since to do so would be documenting performance characteristics that could frequently change or regress. E.g. I see this comment in `SparseCUDATensorMath.cu`:
```
  // FIXME: add some heuristic about when to call coalesce() here, so that
  // tensors don't totally blow up in size by concatenation; e.g.
  //   if (r->nnz / r->minUnique > COMPACTION_THRESHOLD) {
  //     THCSTensor_(contiguous)(r);
  //     r->minUnique = r->nnz;
  //   }
```
If we try to fully document the impact of using uncoalesced sparse tensors in arithmetic ops, that kind of heuristic seems a bit much for user documentation. I understand that users care about performance, but this seems like something where approaches should be explained generally at a high level, or maybe point to blog posts, not necessarily detailing all cases where use of an uncoalesced tensor inside an arithmetic op might lead to additional memory overhead.

Comment (zou3519) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: Yes, you've convinced me that differentiating between coalesced and uncoalesced tensors in the documentation is a bad idea.

However, I still think that we need a note somewhere that in-place sparse division may use more memory than expected -- we could just write "in some cases, sparse in-place division may perform a copy_ of the tensor and not save memory". Just because we have not documented something explicitly doesn't mean that our users don't expect the behavior. As a user of pytorch, I expect in-place operations to not use additional memory.

File: test/test_type_promotion.py [hunk 9]
```diff
@@ -430,6 +431,98 @@ def test_transpose(self, device):
...
+            dense_scalar = op(d1, scalar)
+            self.assertEqual(sparse.dtype, dense_scalar.dtype)
+            self.assertEqual(dense_scalar, sparse.to_dense(), prec=precision, message=err)
+        else:
+            # add(sparse, dense) is not supported. Use add(dense, sparse) instead.
+            # "mul_cpu" / "div_cpu" not implemented for 'Half'
+            self.assertRaises(RuntimeError, lambda: op(s1, d2.view(d2.numel())[0].item()))
+
+    @onlyOnCPUAndCUDA
+    def test_sparse_ops(self, device):
```
Comment (zou3519) [inline test/test_type_promotion.py]: Can this just be `test_sparse_ops(self, device, dtype)`? I think Ruberry's test design allows for it: https://github.com/pytorch/pytorch/blob/3990e9d1ca23e39bfa65fe1d907ddb9f9dbf0919/test/test_torch.py#L6996

File: test/test_type_promotion.py [hunk 10]
```diff
@@ -430,6 +431,98 @@ def test_transpose(self, device):
...
+            sparse = op(s1, scalar)
+            dense_scalar = op(d1, scalar)
+            self.assertEqual(sparse.dtype, dense_scalar.dtype)
+            self.assertEqual(dense_scalar, sparse.to_dense(), prec=precision, message=err)
+        else:
+            # add(sparse, dense) is not supported. Use add(dense, sparse) instead.
+            # "mul_cpu" / "div_cpu" not implemented for 'Half'
+            self.assertRaises(RuntimeError, lambda: op(s1, d2.view(d2.numel())[0].item()))
+
+    @onlyOnCPUAndCUDA
```
Comment (zou3519) [inline test/test_type_promotion.py]: To solve the precision probem, we could use a per-dtype precision override: https://github.com/pytorch/pytorch/blob/3990e9d1ca23e39bfa65fe1d907ddb9f9dbf0919/test/test_torch.py#L13075 .

I know you said that there's a problem with coalesce that means we have to have a custom precision for coalesced ops. We could split `test_sparse_op` in two (have one coalesced-variant and one not-coalesced variant) to work around this.

File: test/test_type_promotion.py (hunk 9 above)
Comment (nairbv) [reply test/test_type_promotion.py]: sort-of, it would support it with a `@dtypes(...a very long list...)`. I don't think we have a way to generically add all math types (and the list of all math types could change). 

Also in the next diff I'm adding a second dtype for mixed-types -- I think it'd be confusing to have one dtype come through annotation but the second come via a for loop.

File: test/test_type_promotion.py (hunk 8 above)
Comment (zou3519) [reply test/test_type_promotion.py]: > If I landed the narrow test just to check the one specific bug I discovered with these tests, I'd kind of prefer to remove that no-longer-needed test once we had complete coverage in the next diff anyway.

Yeah, I agree that would be a lot of movement for no real benefit. Let's just stick with the current tests in this PR then.

> I find this a bit confusing, so I think this might just be a stylistic difference. My main concern with this approach though would be having the list of tests hard-coded, since it leaves a lot of room for missing a case. 

I agree there are pros and cons with each case and I am fine with closing this thread with a "this is a difference in styles" conclusion

File: test/test_type_promotion.py [hunk 11]
```diff
@@ -172,18 +172,19 @@ def test_mixed_type_backward(self, device):
                 func(x, y).sum().backward()
 
     def _get_test_tensor(self, device, dtype, remove_zeros=False):
-        shape = [20, 20, 20]
+        shape = [10, 10, 10]
         if dtype == torch.bool:
-            tensor = torch.randint(0, 2, shape, device=device, dtype=dtype)
+            tensor = torch.randint(int(remove_zeros), 2, shape, device=device, dtype=dtype)
```
Comment (zou3519) [inline test/test_type_promotion.py]: I found this really recently:

https://github.com/pytorch/pytorch/blob/3990e9d1ca23e39bfa65fe1d907ddb9f9dbf0919/test/test_torch.py#L14045-L14066

Could we reuse this? I imagine if we wanted to remove zeros we'd just use `fill_ones=True` or something.

File: test/test_type_promotion.py [hunk 12]
```diff
@@ -430,6 +431,98 @@ def test_transpose(self, device):
         a = torch.tensor([[True, True], [False, True]], device=device)
         self.assertEqual(a.t() == 0, a.t() == False)  # noqa: E712
 
+    def _test_sparse_op_input_tensors(self, device, dtype, coalesced, zeros=True):
```
Comment (zou3519) [inline test/test_type_promotion.py]: The problem with this function is that, eventually,  we are performing binary ops on tensors that are not-very-sparse. It might be good to generate tensors that are somewhat sparse (50%? 75%?) so that we can comprehensively say that our sparse ops are correct.

File: test/test_type_promotion.py (hunk 9 above)
Comment (zou3519) [reply test/test_type_promotion.py]: Oh right, I keep on forgetting the other diff is stacked on top of this one.

File: test/test_type_promotion.py (hunk 10 above)
Comment (nairbv) [reply test/test_type_promotion.py]: it looks like that's tied to the @dtypes annotation, not sure it'd apply here. Also I can only find a single existing use-case for it.

**Review by zou3519 — COMMENTED** [review]
I'm not convinced that the design for testing is the best design for testing for the next diff in the stack. Since I understand how the uncoalesced division fix works (thank you for splitting the PRs apart, I wasn't sure how much of the original PR was fixing this behavior), I think to more effectively review the PR, it might be better to squash the two commits together and review them in one piece to prevent going back and forth between them. Let me know what you think

File: test/test_type_promotion.py (hunk 11 above)
Comment (nairbv) [reply test/test_type_promotion.py]: I'm a bit confused by that function: for half it always returns float, which would then have to be converted externally back to half for the test? I think I'd end up needing a wrapper around it so not sure it would save anything.

fill_ones gives all ones -- to confirm the math is correct I think it's better to have a random/variety of values.

**Comment by nairbv** [comment]
> might be better to squash the two commits together and review them in one piece

I'm fine either way, though if you think you're close to being able to approve this one it'd be nice to get it landed to add current coverage and fix what's broken on master. Also is there an easy way to squash two PRs?

File: test/test_type_promotion.py (hunk 11 above)
Comment (zou3519) [reply test/test_type_promotion.py]: > I'm a bit confused by that function: for half it always returns float, which would then have to be converted externally back to half for the test? I think I'd end up needing a wrapper around it so not sure it would save anything.

A wrapper around it would save 10 LOC and consolidate the testing. I think it is valuable to just have "one way" to perform testing.

> fill_ones gives all ones -- to confirm the math is correct I think it's better to have a random/variety of values.

We could change add a `has_zeros` function to `_make_tensor`: I think that would make sense. The functions in test_torch that call it have `has_zeros` flags. Eventually if we add more integer kernels for other operators, `test_torch` would also benefit from this change.

**Comment by zou3519** [comment]
> I'm fine either way, though if you think you're close to being able to approve this one it'd be nice to get it landed to add current coverage and fix what's broken on master.

I'm not close to being able to approve this one because I'm not sure that the design of the tests in this PR is right for the type promotion PR and I think it is important to be able to look at the whole picture since these two PRs share some common ground.

> Also is there an easy way to squash two PRs?

You can:
- `git rebase -i HEAD~2` (interactive rebase)
- (inside interactive rebase) squash one of the commits into the other
- run `ghstack` again. ghstack should only update one of the pull requests, at which point you can close the other one.

File: test/test_type_promotion.py (hunk 12 above)
Comment (nairbv) [reply test/test_type_promotion.py]: I agree it'd be preferable to have more sparsity. I could probably use a random mask.

**Comment by nairbv** [comment]
replaced by: https://github.com/pytorch/pytorch/pull/30429
